master/internal/cache/cache.go

```diff
@@ -0,0 +1,199 @@
+package cache
+
+import (
+	"context"
+	"io/fs"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/determined-ai/determined/master/internal/db"
+	"github.com/determined-ai/determined/master/pkg/archive"
+	"github.com/determined-ai/determined/proto/pkg/experimentv1"
+)
+
+type modelDefFolder struct {
+	fileTree   []*experimentv1.FileNode
```

Comment:
This struct seems like it should be a map of strings to FileNodes.  I think `insertToTree` would be a lot more performant that way.

Reply:
Using map would definitely help with random accessing, but here I think it makes sense to persist the order

Reply:
Why is that?  Files in directories do not have a natural order.

Reply:
ping

Reply:
From my point of view, tar file decodes to a list, and the API response is also a list, so it's easier to use list in between. Also I was under the impression that when tar file decodes, it persist some sort of order.

Reply:
yes, but I see no real value in the API matching the list order of the tarball.  This code is pretty high algorithmic complexity for no real benefit.

Also I would argue that the protobuf should use a map anyway.

Reply:
I think protobuf should stay using list because map is not compatible with the frontend file tree display frame work we use, otherwise frontend will have to somehow transfer the response to a list

Reply:
Not sure if this is related, but persist some sort of order is important because when user refresh page, we don't want folders/files changing their order

Reply:
> not compatible with the frontend file tree display frame work we use
Is it our code that does the display?  If so, that would be a fine change IMO.

> By "high algorithmic complexity", did you mean what was happening in insertToTree iteration?

Yes.

I think an intermediate solution would be to use a map when building it, but convert it to the list at the end.

Or maybe it's fine as-is.  Probably what I am proposing is too complex.